ext/fast_osc/fast_osc_wrapper.c

```diff
@@ -68,6 +71,45 @@ uint64_t ruby_time_to_osc_timetag(VALUE rubytime) {
   return timetag;
 }
 
+VALUE osc_timetag_to_ruby_time(uint64_t timetag) {
+  uint32_t secs = timetag >> 32;
+  uint32_t frac = timetag;
+  double d_frac = 1.0 * frac / 4294967296.0;
+
+  VALUE c_time = rb_const_get(rb_cObject, rb_intern("Time"));
+  VALUE rb_time = rb_funcall(c_time, rb_intern("at"), 1, DBL2NUM((secs - JAN_1970 + d_frac)));
+
+  return rb_time;
+}
+
+VALUE method_fast_osc_is_bundle(VALUE self, VALUE msg) {
+  char* data = StringValuePtr(msg);
+  return (rtosc_bundle_p(data) ? Qtrue : Qfalse);
+}
+
+VALUE method_fast_osc_decode_bundle(VALUE self, VALUE msg) {
+  char* data = StringValuePtr(msg);
+  VALUE output = rb_ary_new();
+  VALUE args_output = rb_ary_new();
+  // Check that this is indeed a bundle
+  if (!rtosc_bundle_p(data)){
+    return Qnil;
+  }
+
+  uint64_t timetag = rtosc_bundle_timetag(data);
+  rb_ary_push(output, osc_timetag_to_ruby_time(timetag));
+  int n_messages = rtosc_bundle_elements(data, 1024);
+  for (int i = 0; i < n_messages; i++){
+    int message_size = rtosc_bundle_size(data, i+1); // I don't know why, but this function seems to work in 1-index...
+    const char *message = rtosc_bundle_fetch(data, i);
+    VALUE rb_message = rb_str_new(message, message_size);
+    VALUE element = method_fast_osc_decode_single_message(self, rb_message);
```

Comment:
I think this should check if the message is a bundle, and if it is, recurse and merge the contained messages. Note that a bundle containing another bundle may have different timetags, IIUC this will not work with the current API.

Reply:
I totally agree, this is handling only bundling at 1-level. I thought I wold make something quick to solve the current use case, and then implement something more complex.

If we are going to fix it properly, we should also think about changing the output of the functions, so that they can communicate the data that you mention. One possibility, to avoid having recursive lists at the output, is to output a list of (timestamp, path, values) for the decode_bundle function (not just the timestamp at the beginning and then a list of (path, values)).

I can look at implementing that.

Reply:
I agree, and I concur with the API you're suggesting.

I actually had started typing a reply to the original issue thread about API changes with much the same idea: the API for callers should be "parse this message and tell me what OSC commands to run and when". Whether a message is a bundle or not should be opaque to callers (so callers can't accidentally not implement bundle support), and dropping future-dated OSC commands vs scheduling them to run should be an explicit decision from the caller. I believe this would support the entire OSC 1.0 semantics for bundles (and it is compatible with the relaxed semantics for timetags in 1.1).

Thanks a lot for looking into this!

Reply:
Yes, I like the idea that from outside you don't even care if it is a bundle or a discrete message. @xavriley, what do you think?
I think this would mean that instead of having a _decode_single_message_, and a _decode_bundle_, we would just have a _decode_, that processes the data regardless of it being a bundle or a message, and returns a list of (timestamp, path, data). Of course, for messages, the list would be just one element, and with the timestamp set to 1, which I understand is "now" in OSC (or perhaps some other value).